task_02/nikolai_chernyshov/main.cpp

```diff
@@ -0,0 +1,175 @@
+#include <memory>
+#include <cstdlib>
+#include <iostream>
+#include <fstream>
+
+struct edge {
+    static int edge_id_max;
+    int edge_id;
+    int to_vertice;
+    std::shared_ptr<struct vertice> ptr;
+
+    edge() {}
+
+    edge(int to_vertice_, bool again) {
+        if (again) {
+            edge_id = edge_id_max - 1;
+        }
+        else {
+            edge_id = edge_id_max;
+            edge_id_max++;
+        }
+        to_vertice = to_vertice_;
+    }
+};
+
+int edge::edge_id_max = 0;
+
+struct vertice {
+    static int vertice_id_max;
+    int id;
+    int depth;
+    std::unique_ptr<struct edge[]> edges_from_ptr;
+    std::unique_ptr<struct edge> parent_edge_ptr;
+    int edges_from_num;
+
+    //How many new edges to generate from current vertice? 0..3
+    int num_generate() {
+        int res;
+        if (this->depth == 1) return 3;
+        int x = std::rand()/((RAND_MAX + 1u)/10);
+        if ((this->depth == 2) && (x < 2)) return x + 2;
+        if (x <= this->depth) return 0;
+        int y = std::rand()/((RAND_MAX + 1u)/3);
+        if (y == 2) res = 3;
+        else if (y == 1) res = 2;
+        else res = 1;
+        return res;
+    }
+    
+    vertice(int depth_, int parent_id) {
+        id = vertice_id_max;
+        vertice_id_max++;
+        depth = depth_;
+
+        if (depth == 1) {
+            parent_edge_ptr.reset(nullptr);
+        }
+        else {
+            parent_edge_ptr.reset(new struct edge(parent_id, true));
+        }
+        edges_from_num = num_generate();
+
+        if (edges_from_num == 0) {
+            edges_from_ptr.reset(nullptr);
+        }
+        else {
+            edges_from_ptr.reset(new struct edge[edges_from_num]);
+            for (int i=0; i < edges_from_num; i++) {
+                edges_from_ptr[i] = edge(vertice_id_max, false);
+                std::shared_ptr<struct vertice> new_vertice(
+                        new struct vertice(depth + 1, id)
+                    );
+                edges_from_ptr[i].ptr = new_vertice;
+            }
+        }
+    }
+
+    void to_json_vert(std::ofstream& file) {
```

Comment:
Работа с файлами не входит в зону отвественности графа.
Эта функция должна возвращать строку, которая извне уже может быть записана в файл.
```suggestion
    std::string to_json_vert() const {
```

Reply:
Если функция не мутирует родительский объект, то она должна быть помечена `const`
```suggestion
    void to_json_vert(std::ofstream& file) const {
```